File: DSA/Stack/python/minStack.py
class MinStack:
    def __init__(self) -> None:
        self.n=int(input("Enter the length: "))
        self.stack=[]*self.n
        self.min=[]*self.n
        self.stacktop=-1
        self.mintop=-1
        self.run()

    def stackpush(self,num):
        if(self.stacktop==self.n-1):
            print("Overflow Condition Raised...")
        else:
            self.stacktop+=1
            self.stack.append(num)
    def minpush(self,num):
        if(self.mintop==self.n-1):
            print("Overflow Condition Raised...")
        else:
            self.mintop+=1
            self.min.append(num)
    def stackpop(self)->int:
        if(self.stacktop==-1):
            print("Stack is empty, Nothing to POP")
        else:
            temp = self.stack.pop()
            self.stacktop-=1
            return temp
    def minpop(self)->int:
        if(self.mintop==-1):
            print("Stack is empty, Nothing to POP")
        else:
            temp = self.min.pop()
            self.mintop-=1
            return temp
    def stackpeek(self)->int:
        if(self.stacktop==-1):
            print("Stack is empty, Nothing to PEEK")
        else:
            temp=self.stack.pop()
            self.stack.append(temp)
            return temp
    def minpeek(self)->int:
        if(self.mintop==-1):
            print("Stack is empty, Nothing to PEEK")
        else:
            temp=self.min.pop()
            self.min.append(temp)
            return temp
    def run(self):
        flag=True
        while(flag):
            choice=int(input("Select the choice:1.push 2.pop 3.peek 4.MinimumFind 5.exit\n"))
            if(choice==1):
                value=int(input("Enter the number you want to push: "))
                self.stackpush(value)
                if(self.mintop==-1):
                    self.minpush(value)
                elif(self.mintop!=-1 and value<=self.minpeek()):
                    self.minpush(value)
                else:
                    self.minpush(self.minpeek())
            elif(choice==2):
                print(self.stackpop(),"is popped from stack...")
                print(self.minpop(),"is popped from MinStack...")
            elif(choice==3):
                print(self.stackpeek(),"is at the Top of the stack...")
            elif(choice==4):
                print(self.minpeek(),"is the minimum element in the stack...")    
            elif(choice==5):
                flag=False
            else:
                print("Enter Valid choice...")
                
option=MinStack()
    

File: DSA/Stack/C/minStack.c
#include<stdio.h>
#define n 20
int stack[n];
int min[n];
int top=-1;
int mintop=-1;
int isEmpty(){
    if(top==-1 && mintop==-1)
        return 1;
    else
        return 0;
}
int isFull(){
    if(top==n-1 && mintop==n-1)
        return 1;
    else
        return 0;
}
void push(int x){
    if(isFull())
        printf("Stack Overflow....\n");
    else{
        top++;
        stack[top]=x;
    }    
        
}
void pop(){
    if(isEmpty())
        printf("Stack is Empty....\n");
    else
        stack[top--];
}

void minpush(int num){
    if(isFull())
        printf("MinStack Overflow....\n");
    else{
        mintop++;
        min[mintop]=num;
    }

}
void minpop(){
    if(isEmpty())
        printf("MinStack is Empty....\n");
    else
        min[mintop--];
}
int minpeek(){
    if(isEmpty()){
        printf("Stack is Empty....\n");
        return -1;
    }
    else
        return min[mintop];
}
void display(){
    for (int i = 0; i < mintop; i++)
    {
        printf("%d ",min[i]);
    }
    
}
int main(){
    int num,choice;
    int flag=1;
    while(flag){
        printf("1.push 2.pop 3.Min 4.Exit 5.display\nEnter the Choice: ");
        scanf("%d",&choice);
        switch (choice)
        {
        case 1:
                printf("\nEnter the number: ");
                scanf("%d",&num);
                push(num);
                if(mintop==-1 || num<minpeek())
                    minpush(num);
                else
                    minpush(minpeek());
                break;
        case 2:
                pop();
                minpop();
                break;
        case 3:
                printf("%d is Minimum element in the Stack.\n",minpeek());
                break;
        case 4:
                flag=0;
                break;
        case 5:
                display();
                break;
        default:
                printf("Enter valid Choice...\n");
                break;
        }
    }
    return 0;
}